src/advanced/JupyterNotebook/index.tsx

```diff
@@ -0,0 +1,322 @@
+            {children}
+        </SyntaxHighlighter>
+    );
+};
+
+// -- renders a cell of type 'code'
+const JupyterNotebookCodeCell: React.FC<JupyterNotebookCellProps> = ({ cell, executionCount, language }) => {
+    return (
+        <>
+            {cell.metadata && cell.metadata.jupyter && cell.metadata.jupyter.source_hidden ? (
+                <></>
+            ) : (
+                <JupyterNotebookCellRenderTemplate
+                    idColumnContent={`In [${executionCount}]`}
+                    bodyColumnContent={
+                        <ReactMarkdown
+                            components={{ code: CodeBlock }}
+                            children={'```' + language + '\n\n' + sourceLines(cell.source) + '\n```'}
+                        />
+                    }
+                    bodyColumnStyle={{
+                        padding: '0px 0px 0px 0px',
+                        margin: '0px 0px 0px 0px',
+                        border: '1px solid #AAA',
+                    }}
+                />
+            )}
+            {cell.metadata && cell.metadata.jupyter && cell.metadata.jupyter.outputs_hidden ? (
+                <></>
+            ) : (
+                cell.outputs.map((output: any, outputIdx: number) => {
+                    if (output.output_type === 'stream') {
+                        return (
+                            <JupyterNotebookOutputStreamCell
+                                key={`stream-${executionCount}-${outputIdx}`}
+                                cell={cell}
+                                executionCount={executionCount}
+                                output={output}
+                                language={language}
+                            />
+                        );
+                    } else if (output.output_type === 'execute_result' || output.output_type === 'display_data') {
+                        return (
+                            <JupyterNotebookOutputDataCell
+                                key={`exec-${executionCount}-${outputIdx}`}
+                                cell={cell}
+                                executionCount={executionCount}
+                                output={output}
+                                language={language}
+                            />
+                        );
+                    } else if (output.output_type === 'error') {
+                        return (
+                            <JupyterNotebookOutputErrorCell
+                                key={`error-${executionCount}-${outputIdx}`}
+                                cell={cell}
+                                executionCount={executionCount}
+                                output={output}
+                                language={language}
+                            />
+                        );
+                    } else {
+                        return <></>;
+                    }
+                })
+            )}
+        </>
+    );
```

Comment:
it will be better to move this in separate function out of `render`

Reply:
in future refactor